File: leetCode/14.LongestCommonPrefix/LongestCommonPrefix.cpp
#include <string>
#include <vector>

using namespace std;

class TrieNode {
public:
    TrieNode* children[26];
    bool isEnd;
    int linkCount;

    TrieNode() : isEnd(false), linkCount(0) {
        for (int i = 0; i < 26; ++i) {
            children[i] = nullptr;
        }
    }

    void put(char ch, TrieNode* node) {
        if (children[ch - 'a'] == nullptr) {
            children[ch - 'a'] = node;
            linkCount++;
        }
    }

    bool contains(char ch) { return children[ch - 'a'] != nullptr; }

    int getLinks() const { return linkCount; }
};

class Trie {
public:
    TrieNode* root;

    Trie() { root = new TrieNode(); }

    void insert(string word) {
        TrieNode* node = root;
        for (char c : word) {
            if (!node->contains(c)) {
                node->put(c, new TrieNode());
            }
            node = node->children[c - 'a'];
        }
        node->isEnd = true;
    }

    string searchLongestPrefix(string word) {
        TrieNode* node = root;
        string prefix;
        for (char c : word) {
            if (node->contains(c) && node->getLinks() == 1 && !node->isEnd) {
                prefix += c;
                node = node->children[c - 'a'];
            } else {
                break;
            }
        }
        return prefix;
    }
};

class Solution {
public:
    string longestCommonPrefix(string q, vector<string>& strs) {
        if (strs.empty()) return "";
        if (strs.size() == 1) return strs[0];

        Trie trie;
        for (int i = 1; i < strs.size(); i++) {
            trie.insert(strs[i]);
        }
        return trie.searchLongestPrefix(q);
    }
};

File: leetCode/14.LongestCommonPrefix/LongestCommonPrefix.java
class Solution {
    public String longestCommonPrefix(String q, String[] strs) {
        if (strs == null || strs.length == 0) return "";
        if (strs.length == 1) return strs[0];
        Trie trie = new Trie();
        for (int i = 1; i < strs.length; i++) {
            trie.insert(strs[i]);
        }
        return trie.searchLongestPrefix(q);
    }
}

class TrieNode {
    private TrieNode[] links;

    private final int R = 26;

    private boolean isEnd;

    // number of children non null links
    private int size;

    public void put(char ch, TrieNode node) {
        links[ch - 'a'] = node;
        size++;
    }

    public int getLinks() {
        return size;
    }

    public TrieNode() {
        links = new TrieNode[R];
    }

    public boolean containsKey(char ch) {
        return links[ch - 'a'] != null;
    }

    public TrieNode get(char ch) {
        return links[ch - 'a'];
    }

    public void setEnd() {
        isEnd = true;
    }

    public boolean isEnd() {
        return isEnd;
    }
}

class Trie {
    private TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    public String searchLongestPrefix(String word) {
        TrieNode node = root;
        StringBuilder prefix = new StringBuilder();
        for (int i = 0; i < word.length(); i++) {
            char curLetter = word.charAt(i);
            if (
                node.containsKey(curLetter) &&
                node.getLinks() == 1 &&
                !node.isEnd()
            ) {
                prefix.append(curLetter);
                node = node.get(curLetter);
            } else return prefix.toString();
        }
        return prefix.toString();
    }

    public void insert(String word) {
        TrieNode node = root;
        for (int i = 0; i < word.length(); i++) {
            char currentChar = word.charAt(i);
            if (!node.containsKey(currentChar)) {
                node.put(currentChar, new TrieNode());
            }
            node = node.get(currentChar);
        }
        node.setEnd();
    }
}